pip/pip-322.md

```diff
@@ -0,0 +1,143 @@
+
+## In Scope
+
+- Preserve current functionality without breaking changes
+- Consolidate the multiple existing rate limiting options into a single,
+  configurable rate limiting solution
+- Remove the separate “precise” rate limiter
+
+### Problems to Address  
+
+- High CPU load with default rate limiter
+- High lock contention that impacts shared Netty IO threads and adds
+  latency to unrelated Pulsar topic producers ([Bug] RateLimiter lock
+  contention when use precise publish rate limiter #21442.)
+- Multiple limiting implementations (default, precise) which
+  unnecessarily expose implementation details to users of Pulsar and
+  make the code harder to maintain and improve
+- Inability to limit throughput consistently when using default rate
+  limiter
+- Inconsistent behavior across multiple levels of throttling (broker,
+  namespace, connection)
+- Code maintainability
+  - Improve understandability of code
+
+## Out of Scope
+
+- Custom/pluggable rate limiters
+- Additional rate limiting features
+- Cluster-wide capacity management
+- Addressing the shared connection multiplexing problem where throttling
+  multiple independent streams multiplexed on the same connection cannot
+  be consistently throttled by pausing reads on the server side.
+
+# High Level Design
+
+The proposed refactor will overhaul rate limiting internals while
+preserving existing APIs and user-facing behavior. A token bucket
+algorithm will provide efficient and accurate calculations to throttle
+throughput. 
+
+Multiple built-in options such as "precise" rate limiter will be
+consolidated under a single solution. Performance issues caused by
+contention and CPU overhead will be addressed.
+
+## Publisher rate limiter throttling
+
+The refactored design ensures that a publish rate limiter operates in a
+non-blocking fashion. It consumes the quota seamlessly without
+disrupting ongoing operations. When the quota is exhausted, the
+connection is throttled by pausing reads.
+
+Subsequently, the publisher rate limiter adds the throttled producer to
+a queue for unthrottling and schedules a task to process this queue
+after a certain delay, assuming there isn't a task already scheduled.
+When this task is executed, it unthrottles all the publishers in the
+queue, provided there is available quota. This queuing mechanism
+guarantees fairness in the solution.
+
+In terms of implementation, the rate limiter necessitates a token bucket
+algorithm that is asynchronous, lockless, and non-blocking. This
+algorithm should be capable of calculating the duration of throttling. A
+proof-of-concept implementation that fulfills these requirements can be
+found at https://github.com/lhotari/async-tokenbucket.
+
+# Detailed Design
+
+## Proposed Solution
+
+- Use asynchronous token bucket algorithm
+  - Efficient for concurrent updates
+    - Can be implemented in a lockless, non-blocking way using CAS
+      fields. 
+  - Conceptually simple
+  - Common traffic shaping approach   
```

Comment:
You mean future development can modify the algorithm to introduce traffic shaping?